oscore-cf/src/main/java/org/eclipse/californium/oscore/OSSerializer.java

```diff
@@ -0,0 +1,557 @@
+/*******************************************************************************
+ * Copyright (c) 2018 RISE SICS and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Joakim Brorsson
+ *    Ludwig Seitz (RISE SICS)
+ *    Tobias Andersson (RISE SICS)
+ *    
+ ******************************************************************************/
+package org.eclipse.californium.oscore;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Logger;
+
+import org.eclipse.californium.core.coap.CoAP;
+import org.eclipse.californium.core.coap.Option;
+import org.eclipse.californium.core.coap.OptionSet;
+import org.eclipse.californium.elements.util.DatagramReader;
+import org.eclipse.californium.elements.util.DatagramWriter;
+
+import com.upokecenter.cbor.CBORObject;
+
+public class OSSerializer {
+
+	/**
+	 * The logger
+	 */
+	private static final Logger LOGGER = Logger.getLogger(OSSerializer.class.getName());
+
+	/**
+	 * Prepare options and payload for encrypting.
+	 * 
+	 * @param options the options
+	 * 
+	 * @param payload the payload
+	 * 
+	 * @return the serialized plaintext for OSCore
+	 */
+	public static byte[] serializeConfidentialData(OptionSet options, byte[] payload, int realCode) {
+		if (options != null) {
+			if (realCode > 0) {
+				DatagramWriter writer = new DatagramWriter();
+				writeOptions(writer, options, true, false, realCode);
+				if (payload != null && payload.length > 0) {
```

Comment:
Already checked in writePayload.

Should we make the `DataSerializer.serializeOptionsAndPayload` public and add the `DatagramWriter` together with an new `OptionFilter` to the parameter for reuse the mostly code copy there?

Reply:
That would also make the code more readable.

Reply:
with PR #660 it should look similar to:

```
OptionSet filteredOptions = OptionJuggle.prepareEoptions(options);
writer.write(realCode, CoAP.MessageFormat.CODE_BITS);
DataSerializer.serializeOptionsAndPayload(writer, filteredOptions, payload);

```

I tried also a `OptionFilter` approach to reduce the "juggle", but the PROXY-URI seems to be too complex.
With that, a lot of cleanup should be possible.